File: src/main/java/nuclearcoder/discordbot/CustomModuleLoader.java
package nuclearcoder.discordbot;

import org.apache.commons.io.filefilter.FileFilterUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import sx.blah.discord.modules.IModule;
import sx.blah.discord.modules.ModuleLoader;

import java.io.File;
import java.io.FilenameFilter;
import java.io.IOException;
import java.net.URL;
import java.net.URLClassLoader;
import java.util.jar.JarFile;
import java.util.jar.Manifest;

public class CustomModuleLoader {

    private static final File MODULES_DIR = new File("modules");
    private static final Logger LOGGER = LoggerFactory.getLogger(CustomModuleLoader.class);

    private static final File[] listModulesDirectory()
    {
        return MODULES_DIR.listFiles((FilenameFilter) FileFilterUtils.suffixFileFilter("jar"));
    }

    private static Class loadClass(ClassLoader classLoader, String clazz)
            throws ClassNotFoundException
    {
        if (clazz.contains("$") && clazz.substring(0, clazz.lastIndexOf("$")).length() > 0)
        {
            try
            {
                loadClass(classLoader, clazz.substring(0, clazz.lastIndexOf("$")));
            }
            catch (ClassNotFoundException ignored)
            {
                // If the parent class doesn't exist then it is safe to instantiate the child
            }
        }
        return Class.forName(clazz, true, classLoader);
    }

    private static final void loadFile(File file, ModuleLoader loader)
    {
        if (file.isFile() && file.getName().endsWith(".jar"))
        { // Can't be a directory and must be a jar
            try (JarFile jar = new JarFile(file))
            {
                Manifest man = jar.getManifest();
                String moduleAttrib = man.getMainAttributes().getValue("Discord4J-ModuleClass");

                if (moduleAttrib != null)
                {
                    String[] moduleClasses = moduleAttrib.split(";");

                    try (URLClassLoader classLoader = new URLClassLoader(
                            new URL[] { file.toURI().toURL() }))
                    {
                        for (String moduleClass : moduleClasses)
                        {
                            Class classInstance = loadClass(classLoader, moduleClass);
                            if (IModule.class.isAssignableFrom(classInstance))
                            {
                                IModule module = (IModule) classInstance.newInstance();

                                LOGGER.info("Loading module {} v{} by {}", module.getName(),
                                        module.getVersion(), module.getAuthor());

                                loader.loadModule(module);
                            }
                        }
                    }
                }
            }
            catch (IllegalAccessException | InstantiationException | ClassNotFoundException | IOException e)
            {
                LOGGER.error("Unable to load module " + file.getName() + "!", e);
            }
        }
    }

    static final void reloadModules(ModuleLoader loader)
    {
        // unload them
        for (IModule module : loader.getLoadedModules())
            loader.unloadModule(module);

        // load new
        File[] files = listModulesDirectory();
        if (files != null && files.length > 0)
        {
            LOGGER.info("Attempting to (re-)load {} module(s)...", files.length);

            for (File file : files)
                loadFile(file, loader);
        }
    }

}

File: src/main/java/nukebot/CustomModuleLoader.kt
package nukebot

import sx.blah.discord.modules.IModule
import sx.blah.discord.modules.ModuleLoader
import java.io.File
import java.io.IOException
import java.net.URLClassLoader
import java.util.jar.JarFile

private fun listModulesDirectory(): Array<File> {
    return MODULES_DIR.listFiles { file, s -> s.endsWith(".jar", true) } ?: emptyArray()
}

private fun loadClass(classLoader: ClassLoader, clazz: String): Class<*> {
    if (clazz.contains("$")) {
        val subString = clazz.substring(0, clazz.lastIndexOf("$"))
        if (subString.isNotEmpty()) {
            try {
                loadClass(classLoader, subString)
            } catch (ignored: ClassNotFoundException) {
                // If the parent class doesn't exist then it is safe to instantiate the child
            }
        }
    }
    return Class.forName(clazz, true, classLoader)
}

private fun loadFile(file: File, loader: ModuleLoader) {
    if (file.isFile && file.name.endsWith(".jar")) { // Can't be a directory and must be a jar
        try {
            JarFile(file).use { jar ->
                val man = jar.manifest
                val moduleAttrib = man.mainAttributes.getValue("Discord4J-ModuleClass") ?: return

                val moduleClasses = moduleAttrib.split(';')

                URLClassLoader(
                        arrayOf(file.toURI().toURL())).use {
                    for (moduleClass in moduleClasses) {
                        val classInstance = loadClass(it, moduleClass)
                        if (IModule::class.java.isAssignableFrom(classInstance)) {
                            val module = classInstance.newInstance() as IModule

                            LOGGER.info("Loading module {} v{} by {}", module.name,
                                    module.version, module.author)

                            loader.loadModule(module)
                        }
                    }
                }
            }
        } catch (e: Exception) {
            when (e) {
                is IllegalAccessException,
                is InstantiationException,
                is ClassNotFoundException,
                is IOException
                -> LOGGER.error("Unable to load module " + file.name, e)
                else -> throw e
            }
        }

    }
}

internal fun reloadModules(loader: ModuleLoader) {
    // unload them
    loader.loadedModules.forEach { loader.unloadModule(it) }

    // load new
    val files = listModulesDirectory()

    LOGGER.info("Attempting to (re-)load {} module(s)...", files.size)
    if (files.isNotEmpty()) {
        files.forEach { loadFile(it, loader) }
    }
}
